File: share/webapp/js/room/typing.js
/*
// Matrix Construct
//
// Copyright (C) Matrix Construct Developers, Authors & Contributors
// Copyright (C) 2016-2018 Jason Volk <jason@zemos.net>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice is present in all copies. The
// full license for this software is available in the LICENSE file.
*/

'use strict';


/**************************************
 * Typing notifications
 *
 */

room.typing = {};

/**
 * State of mxid => bool
 */
room.typing.active = {};

/**
 * Sends a typing event for my user in this room
 *
 * this = room
 */
room.typing.set = function(bool)
{
	if(this.opts.local)
		return;

	if(this.control.evalmode && bool)
		return;

	let mxid = mc.my.mxid;
	if(this.typing.active[mxid] === bool)
		return;

	let opts =
	{
		content:
		{
			typing: bool,
			timeout: 15 * 1000,
		},
	};

	this.typing.requesting = true;
	this.typing.active[mxid] = bool;
	client.m.rooms.typing.put(this.id, mxid, opts, (error, data) =>
	{
		this.typing.requesting = false;

		if(error)
		{
			delete this.typing.active[mxid];
			throw new client.error(error);
		}

		if(bool)
			this.typing.active[mxid] = true;
		else
			delete this.typing.active[mxid];
	});
};

/**
 * Function for room.typing.active, which is a dictionary of mxid => boolean.
 *
 * this = room
 */
room.typing.get = function(mxid = mc.my.mxid)
{
	return this.typing.active[mxid];
};

/**
 * Receives a list of users currently typing to update the view.
 *
 * this = room
 */
room.typing.update = function(user_ids = [])
{
	Object.each(this.typing.active, (key, val) =>
	{
		if(!user_ids.includes(key))
			delete this.typing.active[key];
	});

	user_ids.forEach((user_id) =>
	{
		this.typing.active[user_id] = true;
	});
};

File: modules/client/sync/rooms/ephemeral/typing.cc
// Matrix Construct
//
// Copyright (C) Matrix Construct Developers, Authors & Contributors
// Copyright (C) 2016-2019 Jason Volk <jason@zemos.net>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice is present in all copies. The
// full license for this software is available in the LICENSE file.

namespace ircd::m::sync
{
	static bool room_ephemeral_m_typing_polylog(data &);
	static bool room_ephemeral_m_typing_linear(data &);
	extern item room_ephemeral_m_typing;
}

ircd::mapi::header
IRCD_MODULE
{
    "Client Sync :Room Ephemeral :Typing"
};

decltype(ircd::m::sync::room_ephemeral_m_typing)
ircd::m::sync::room_ephemeral_m_typing
{
	"rooms.ephemeral.m_typing",
	room_ephemeral_m_typing_polylog,
	room_ephemeral_m_typing_linear
};

bool
ircd::m::sync::room_ephemeral_m_typing_linear(data &data)
{
	if(json::get<"type"_>(*data.event) != "ircd.typing")
		return false;

	const m::room target_room
	{
		unquote(json::get<"content"_>(*data.event).get("room_id"))
	};

	// Check if our user is a member of the room targetted by the typing notif
	if(!membership(target_room, data.user, "join"))
		return false;

	const m::user::id &sender
	{
		json::get<"sender"_>(*data.event)
	};

	const m::user::room user_room
	{
		sender
	};

	// Check if the ircd.typing event was sent to the user's user room,
	// and not just to any room.
	if(json::get<"room_id"_>(*data.event) != user_room.room_id)
		return false;

	// Check if the user does not want to receive typing events for this room.
	if(!m::typing::allow(data.user, *data.room, "sync"))
		return false;

	json::stack::object rooms
	{
		*data.out, "rooms"
	};

	json::stack::object membership_
	{
		*data.out, "join"
	};

	json::stack::object room_
	{
		*data.out, target_room.room_id
	};

	json::stack::object ephemeral
	{
		*data.out, "ephemeral"
	};

	json::stack::array events
	{
		*data.out, "events"
	};

	json::stack::object object
	{
		*data.out
	};

	json::stack::member
	{
		object, "type", "m.typing"
	};

	json::stack::object content
	{
		object, "content"
	};

	json::stack::array user_ids
	{
		content, "user_ids"
	};

	m::typing::for_each([&user_ids, &target_room]
	(const auto &typing)
	{
		if(json::get<"room_id"_>(typing) != target_room)
			return true;

		user_ids.append(json::get<"user_id"_>(typing));
		return true;
	});

	return true;
}

bool
ircd::m::sync::room_ephemeral_m_typing_polylog(data &data)
{
	return false;
}
